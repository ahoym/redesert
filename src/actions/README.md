# Redesert Action Types

`resourceApiActionTypesFactory` creates action types that
`redesert-resource-reducer`s automatically have cases for, and
`redesert-api-thunk`s automatically dispatches.

In general, they look like:

```js
`${API_ACTION_PREFIX}__${anyResource}__${API_LIFECYCLE_SUFFIX}`;
```

# API Action Prefixes

These are meant to mirror REST actions towards a service endpoint. In general:

* **REST Action** - **`redesert-api-action-type`**
* `GET` - `FETCH__*`
* `POST` - `CREATE-__*`
* `PUT`/`PATCH` - `UPDATE__*`
* `DELETE` - `REMOVE__*`

These will let us create consistently prefixed action types, no matter what
the resource (\*) is. This allows us to predictably know when and what
API action is broadcasted, and create general state transformations based on
specific REST actions. This is how we can re-use the same `fetch*`, `create*`,
`update*`, and `remove*` reducers no matter what the resource is.

To note, `redesert-resource-reducer`s initially guard against actions that do
not contain the `resource` name within the type, which is how the redcuers know
how to distinguish which actions to act upon or not.

# API Lifecycle Suffixes

All network requests have a basic logic flow of:

* `*__START` a network request. Signifies the request is in flight
* `*__SUCCESS` response from the endpoint (success)
* `*__FAILURE` response from the endpoint (error)

Of course there can be other types of suffixes (maybe `*__IN_PROGRESS` for
example). But in general the above three is the assumption, no matter what type
of API action is dispatched.

Because we have predictable API lifecycle suffixes, we can do two things:

1.  Create a light wrapper around our network requesting modules that
    auto-dispatches these lifecycle actions (`redesert-api-thunk`)
2.  Within reducers, we can listen to these life cycles and perform general state
    transformations to during each action of an API request lifecycle

Both can be resource agnostic. `redesert-api-thunk` will accept a
`baseActionType`, which contains information about the API action type and the
resource. It will then use that base type and dispatch suffixed variants, based
on where the network request is in the lifecycle. On the other side of things,
`redesert-resource-reducers` have cases for these suffixes in each API reducer
and performs general state updates based on the status of the API request.
