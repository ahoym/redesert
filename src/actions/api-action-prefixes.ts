export interface ApiActionPrefixes {
  readonly FETCH: 'FETCH_';
  readonly CREATE: 'CREATE_';
  readonly UPDATE: 'UPDATE_';
  readonly REMOVE: 'REMOVE_';
}

const API_ACTION_PREFIXES: ApiActionPrefixes = {
  FETCH: 'FETCH_',
  CREATE: 'CREATE_',
  UPDATE: 'UPDATE_',
  REMOVE: 'REMOVE_',
};

export function resourceApiActionTypesFactory(resource: string) {
  return {
    FETCH: `${API_ACTION_PREFIXES.FETCH}_${resource}`,
    CREATE: `${API_ACTION_PREFIXES.CREATE}_${resource}`,
    UPDATE: `${API_ACTION_PREFIXES.UPDATE}_${resource}`,
    REMOVE: `${API_ACTION_PREFIXES.REMOVE}_${resource}`,
  };
}

export default API_ACTION_PREFIXES;
