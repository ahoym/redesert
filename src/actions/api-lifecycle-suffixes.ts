export interface ApiLifeCycleSuffixes {
  readonly START: '_START';
  readonly SUCCESS: '_SUCCESS';
  readonly FAILURE: '_FAILURE';
}

const API_LIFECYCLE_SUFFIXES: ApiLifeCycleSuffixes = {
  START: '_START',
  SUCCESS: '_SUCCESS',
  FAILURE: '_FAILURE',
};

export const ALL_API_LIFECYCLE_SUFFIXES: string[] = Object.keys(
  API_LIFECYCLE_SUFFIXES
);

export default API_LIFECYCLE_SUFFIXES;
