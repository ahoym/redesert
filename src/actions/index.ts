export { default as API_LIFECYCLE_SUFFIXES } from './api-lifecycle-suffixes';
export { resourceApiActionTypesFactory } from './api-action-prefixes';
export { default as API_ACTION_PREFIXES } from './api-action-prefixes';
export { default as isValidActionType } from './is-valid-action-type';
export * from './type-definitions';
