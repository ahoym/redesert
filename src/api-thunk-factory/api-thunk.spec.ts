import apiThunk from './api-thunk';
import { resourceApiActionTypesFactory } from '../actions';

describe('apiThunk()', () => {
  const mockResponse = {
    id: '123',
    name: 'foo',
  };
  const dispatchFn = jest.fn();
  const fooApiActions = resourceApiActionTypesFactory('foo');
  const baseActionType = fooApiActions.FETCH;
  const networkRequest = jest
    .fn()
    .mockReturnValue(Promise.resolve(mockResponse));
  const defaultProps = {
    baseActionType,
    networkRequest,
  };

  const createApiThunkAndInvoke = (thunkProps: object = {}) => {
    const actionCreator = apiThunk({ ...defaultProps, ...thunkProps });
    return actionCreator(dispatchFn);
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('successful networkRequest responses', () => {
    it('dispatches the *__START and *__SUCCESS actions', async () => {
      await createApiThunkAndInvoke();
      expect(dispatchFn.mock.calls).toMatchSnapshot();
    });

    it('dispatches the actions with meta attributes', async () => {
      const meta = { referenceId: '123' };
      await createApiThunkAndInvoke({ meta });
      expect(dispatchFn.mock.calls).toMatchSnapshot();
    });

    it('uses the successNormalizer to parse the response for the action.payload', async () => {
      const successNormalizer = (mockResp: object) => ({
        ...mockResponse,
        added: 'from success normalizer',
      });
      const normalizedResp = await createApiThunkAndInvoke({
        successNormalizer,
      });

      expect(normalizedResp).toEqual({
        ...mockResponse,
        added: 'from success normalizer',
      });
    });

    it('returns a Promise-y value that also can be chained off of', () => {
      return createApiThunkAndInvoke().then((resp: any) => {
        expect(resp).toEqual(mockResponse);
      });
    });
  });

  describe('failed networkRequest responses', () => {
    const mockFailureResponse = {
      status: 500,
      statusText: 'Internal server error',
    };

    beforeAll(() => {
      networkRequest.mockRejectedValue(mockFailureResponse);
    });

    it('dispatches the *__START and *__ERROR actions', async () => {
      try {
        await createApiThunkAndInvoke();
      } catch (error) {
        expect(dispatchFn.mock.calls).toMatchSnapshot();
      }
    });

    it('dispatches the actions with meta attributes', async () => {
      try {
        const meta = { referenceId: '123' };
        await createApiThunkAndInvoke({ meta });
      } catch (error) {
        expect(dispatchFn.mock.calls).toMatchSnapshot();
      }
    });

    it('uses the failureNormalizer to parse the response for the action.errors', async () => {
      try {
        const failureNormalizer = (mockResp: any) => [mockResp.statusText];
        await createApiThunkAndInvoke({ failureNormalizer });
      } catch (error) {
        expect(dispatchFn).toHaveBeenLastCalledWith(
          expect.objectContaining({
            errors: [mockFailureResponse.statusText],
          })
        );
      }
    });

    it('returns a Promise-y value that also can be chained off of', () => {
      return createApiThunkAndInvoke().catch((error: any) => {
        expect(error).toEqual(mockFailureResponse);
      });
    });
  });
});
