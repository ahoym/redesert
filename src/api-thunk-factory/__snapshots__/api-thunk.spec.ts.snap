// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiThunk() failed networkRequest responses dispatches the *__START and *__ERROR actions 1`] = `
Array [
  Array [
    Object {
      "meta": undefined,
      "type": "FETCH__foo__START",
    },
  ],
  Array [
    Object {
      "errors": Object {
        "status": 500,
        "statusText": "Internal server error",
      },
      "meta": undefined,
      "type": "FETCH__foo__FAILURE",
    },
  ],
]
`;

exports[`apiThunk() failed networkRequest responses dispatches the actions with meta attributes 1`] = `
Array [
  Array [
    Object {
      "meta": Object {
        "referenceId": "123",
      },
      "type": "FETCH__foo__START",
    },
  ],
  Array [
    Object {
      "errors": Object {
        "status": 500,
        "statusText": "Internal server error",
      },
      "meta": Object {
        "referenceId": "123",
      },
      "type": "FETCH__foo__FAILURE",
    },
  ],
]
`;

exports[`apiThunk() successful networkRequest responses dispatches the *__START and *__SUCCESS actions 1`] = `
Array [
  Array [
    Object {
      "meta": undefined,
      "type": "FETCH__foo__START",
    },
  ],
  Array [
    Object {
      "meta": undefined,
      "payload": Object {
        "id": "123",
        "name": "foo",
      },
      "type": "FETCH__foo__SUCCESS",
    },
  ],
]
`;

exports[`apiThunk() successful networkRequest responses dispatches the actions with meta attributes 1`] = `
Array [
  Array [
    Object {
      "meta": Object {
        "referenceId": "123",
      },
      "type": "FETCH__foo__START",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "referenceId": "123",
      },
      "payload": Object {
        "id": "123",
        "name": "foo",
      },
      "type": "FETCH__foo__SUCCESS",
    },
  ],
]
`;
